### PR TITLE
feat: Check nodestore instead of Events table for duplicate event

### PR DESCRIPTION
**Pull request by lynnagara**
Since we plan to remove the Events table, we instead check nodestore
before processing an event for the presence of a duplicate ID.

The previously discussed approach of reprocessing all duplicates and
changing from first-write-wins to last-write-wins will not work for
events across different days since Snuba uses "toStartOfDay(timestamp)"
as part of the sorting key.

Ref: SNS-238

### CODE REVIEW

**Review by mattrobenolt — COMMENTED** [review]
I'm not going to block this... but this is going to be an extremely expensive thing to do. We absolutely need to test this behind some switch so we can dial it up/down. This is adding a new GET request to every single event that comes through. So thousands of GETs/s being added.

File: src/sentry/options/defaults.py [hunk 1]
```diff
@@ -167,6 +167,9 @@
 register("store.normalize-after-processing", default=0.0)  # unused
 register("store.disable-trim-in-renormalization", default=0.0)  # unused
 
+# Use nodestore instead of events table for duplicate checking in store
+register("store.use-nodestore", type=Bool, default=True, flags=FLAG_PRIORITIZE_DISK)
```
Comment (lynnagara) [inline src/sentry/options/defaults.py]: this will be False

File: src/sentry/event_manager.py [hunk 2]
```diff
@@ -513,16 +523,14 @@ def save(self, project_id, raw=False, assume_normalized=False):
...
-        # there's a race condition between here and when the event is actually
-        # saved, but it's an improvement. See GH-7677.)
-        try:
-            event = Event.objects.get(project_id=project.id, event_id=data["event_id"])
-        except Event.DoesNotExist:
-            pass
-        else:
+        # isn't a perfect solution -- there's a race condition between here and
+        # when the event is actually saved, but it's an improvement. See GH-7677.)
+        event = self._get_event(project_id, data["event_id"])
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: When we get here we have already excluded duplicate event ids that were in memcache, is that right ?

File: src/sentry/event_manager.py [hunk 3]
```diff
@@ -875,6 +883,19 @@ def save(self, project_id, raw=False, assume_normalized=False):
 
         return event
 
+    def _get_event(self, project_id, event_id):
+        if options.get("store.use-nodestore"):
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: Could we have this flag set to a percentage? So we can try to roll it out gradually and assess the performance impact before we fully rely on nodestore? This would make the rollout safer.

File: src/sentry/event_manager.py [hunk 4]
```diff
@@ -875,6 +883,19 @@ def save(self, project_id, raw=False, assume_normalized=False):
...
+            node_data = nodestore.get(Event.generate_node_id(project_id, event_id))
+            if node_data:
+                return eventstore.get_event_by_id(project_id, event_id)
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: Why do we need to load the event object from snuba when we already have the event content from nodestore ?

File: tests/sentry/event_manager/test_event_manager.py [hunk 5]
```diff
@@ -43,7 +43,7 @@
 
 def make_event(**kwargs):
     result = {
-        "event_id": "a" * 32,
+        "event_id": uuid.uuid1().hex,
```
Comment (fpacifici) [inline tests/sentry/event_manager/test_event_manager.py]: why this change ?

File: src/sentry/event_manager.py [hunk 6]
```diff
@@ -513,16 +523,14 @@ def save(self, project_id, raw=False, assume_normalized=False):
 
         # Check to make sure we're not about to do a bunch of work that's
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: This is not accurate anymore. Our concern is now that the order by clause in Clickhouse include the timestamp of the event (rounded to the beginning of the day) so potentially, two events with the same project_id and event_id and timestamp far apart enough would not be merged by clickhouse thus creating an inconsistency in the data model.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Matt is right, please ensure you sync with ops before testing this.

File: src/sentry/event_manager.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/event_manager.py]: Good point, we should definitely avoid this

File: src/sentry/event_manager.py (hunk 2 above)
Comment (lynnagara) [reply src/sentry/event_manager.py]: Yep, that's right

File: tests/sentry/event_manager/test_event_manager.py (hunk 5 above)
Comment (lynnagara) [reply tests/sentry/event_manager/test_event_manager.py]: We actually do want different events rather than duplicates in these tests

File: src/sentry/event_manager.py (hunk 3 above)
Comment (lynnagara) [reply src/sentry/event_manager.py]: Yes, will update

File: src/sentry/options/defaults.py (hunk 1 above)
Comment (markstory) [reply src/sentry/options/defaults.py]: But the default is True right now?

Comment (lynnagara) [reply src/sentry/options/defaults.py]: Sorry, it was just a note to myself to flip it later

File: src/sentry/event_manager.py [hunk 7]
```diff
@@ -511,16 +512,13 @@ def save(self, project_id, raw=False, assume_normalized=False):
...
-        # saved, but it's an improvement. See GH-7677.)
-        try:
-            event = Event.objects.get(project_id=project.id, event_id=data["event_id"])
-        except Event.DoesNotExist:
-            pass
-        else:
+        # Ensure an event with the same ID does not exist before processing it.
+        # We use a first write wins approach since Clickhouse cannot merge
+        # events from different days. (The timestamp rounded to
+        # start of day is part of the ordering key in Clickhouse).
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: nit: primary key and not ordering key. Here they happen to be the same but in principle they may be different and it is the primary key that drives the merge.

File: src/sentry/event_manager.py [hunk 8]
```diff
@@ -875,6 +873,22 @@ def save(self, project_id, raw=False, assume_normalized=False):
 
         return event
 
+    def _get_event(self, project_id, event_id):
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: nit: let's rename to get_event_from_storage, or load_event or something that specifies that we are fetching the event from the storage. Since we are in event manager, get event may mean that we are getting it from the input and what is critical is where we are getting it from.

File: src/sentry/event_manager.py [hunk 9]
```diff
@@ -875,6 +873,22 @@ def save(self, project_id, raw=False, assume_normalized=False):
...
+        nodestore_sample_rate = options.get("store.nodestore-sample-rate")
+        use_nodestore = random.random() < nodestore_sample_rate
+
+        if use_nodestore:
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: When you discuss with James the rollout process, please ensure you have reliable metrics about:
- how often we return an event here
- how long it takes to load an event when it exists 
- how long it takes to load an event when it does not exist. (The load we are putting on nodestore here is vastly skewed towards reads of elements that do not exist. I am not sure we have a good sense around the performance of this specific load pattern (several systems are a lot less efficient at trying to find an element that does not exist than loading an element whern it exists)).

File: tests/sentry/event_manager/test_event_manager.py (hunk 5 above)
Comment (fpacifici) [reply tests/sentry/event_manager/test_event_manager.py]: you can pass the event_id as parameter. Anyway, if this is not breaking the other tests in the file, that's ok

**Review by fpacifici — COMMENTED** [review]
This is generally ok, please, before this gets accepted, ensure you have a rollout plan with James. Specifically ensure we have all the needed monitoring, then I will accept

File: src/sentry/event_manager.py (hunk 7 above)
Comment (lynnagara) [reply src/sentry/event_manager.py]: Thanks, updated

File: src/sentry/event_manager.py (hunk 8 above)
Comment (lynnagara) [reply src/sentry/event_manager.py]: Done

**Review by fpacifici — APPROVED** [review]
